subprojects/logging/src/main/java/org/gradle/internal/logging/sink/GroupingProgressLogEventGenerator.java

```diff
@@ -0,0 +1,180 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.logging.sink;
+
+import org.gradle.api.Nullable;
+import org.gradle.internal.SystemProperties;
+import org.gradle.internal.logging.events.BatchOutputEventListener;
+import org.gradle.internal.logging.events.EndOutputEvent;
+import org.gradle.internal.logging.events.LogEvent;
+import org.gradle.internal.logging.events.OperationIdentifier;
+import org.gradle.internal.logging.events.OutputEvent;
+import org.gradle.internal.logging.events.OutputEventListener;
+import org.gradle.internal.logging.events.ProgressCompleteEvent;
+import org.gradle.internal.logging.events.ProgressEvent;
+import org.gradle.internal.logging.events.ProgressStartEvent;
+import org.gradle.internal.logging.events.RenderableOutputEvent;
+import org.gradle.internal.logging.events.StyledTextOutputEvent;
+import org.gradle.internal.logging.text.StyledTextOutput;
+import org.gradle.internal.progress.BuildOperationType;
+import org.gradle.internal.time.TimeProvider;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An {@code org.gradle.logging.internal.OutputEventListener} implementation which generates output events to log the
+ * progress of operations.
+ */
+public class GroupingProgressLogEventGenerator extends BatchOutputEventListener {
+    private static final String EOL = SystemProperties.getInstance().getLineSeparator();
+
+    private final OutputEventListener listener;
+    private final TimeProvider timeProvider;
+
+    // Maintain a hierarchy of all build operation ids — heads up: this is a *forest*, not just 1 tree
+    private final Map<Object, Object> buildOpIdHierarchy = new LinkedHashMap<Object, Object>();
+    private final Map<Object, OperationGroup> operationsInProgress = new LinkedHashMap<Object, OperationGroup>();
+    private final Map<OperationIdentifier, Object> progressToBuildOpIdMap = new LinkedHashMap<OperationIdentifier, Object>();
+
+    private Object lastRenderedBuildOpId;
+
+    public GroupingProgressLogEventGenerator(OutputEventListener listener, TimeProvider timeProvider) {
+        this.listener = listener;
+        this.timeProvider = timeProvider;
+    }
+
+    public void onOutput(OutputEvent event) {
+        if (event instanceof ProgressStartEvent) {
+            onStart((ProgressStartEvent) event);
+        } else if (event instanceof RenderableOutputEvent) {
+            handleOutput((RenderableOutputEvent) event);
+        } else if (event instanceof ProgressCompleteEvent) {
+            onComplete((ProgressCompleteEvent) event);
+        } else if (event instanceof EndOutputEvent) {
```

Comment:
One more thing that comes to mind. If an exception is thrown during build execution are we sure that we'll always receive an end output event? If we don't then we should probably clear the Maps through some other mechanism.